source/tit/geom/partition/grid_graph_partition.hpp

```diff
@@ -0,0 +1,149 @@
+  /// @param size_hint Grid cell size, typically 2x of the particle spacing.
+  constexpr explicit GridGraphPartition(
+      real_t size_hint,
+      GraphPartition graph_partition = {}) noexcept
+      : size_hint_{size_hint}, graph_partition_{std::move(graph_partition)} {
+    TIT_ASSERT(size_hint_ > 0.0, "Cell size hint must be positive!");
+  }
+
+  /// Partition the points using the grid graph partitioning algorithm.
+  template<point_range Points, output_index_range Parts>
+  void operator()(Points&& points,
+                  Parts&& parts,
+                  size_t num_parts,
+                  size_t init_part = 0) const {
+    TIT_PROFILE_SECTION("GridGraphPartition::operator()");
+    TIT_ASSUME_UNIVERSAL(Points, points);
+    TIT_ASSUME_UNIVERSAL(Parts, parts);
+    static constexpr auto Dim = point_range_dim_v<Points>;
+
+    // Validate the arguments.
+    TIT_ASSERT(num_parts > 0, "Number of parts must be positive!");
+    TIT_ASSERT(std::size(points) >= num_parts,
+               "Number of points cannot be less than the number of parts!");
+    if constexpr (std::ranges::sized_range<Parts>) {
+      TIT_ASSERT(std::size(points) == std::size(parts),
+                 "Size of parts range must be equal to the number of points!");
+    }
+
+    // Compute bounding box and initialize the grid.
+    // We'll extend the number of cells by one in each direction to avoid the
+    // conditionals near boundary.
+    //
+    // Note: "100" does not mean anything particular, it's just for historical
+    //       reasons, and it can be replaced with any value > 1.
+    const auto box = compute_bbox(points).grow(size_hint_ / 100);
+    const auto grid = Grid{box}.set_cell_extents(size_hint_).extend(1);
+
+    // Index the cells that contain the points. Thouse would be used as nodes
+    // in the graph. We'll use amount of points in each cell as the node weight.
+    //
+    // Since the typical SPH adjacency graph is heavily connected, we'll use
+    // the product of the node weights as the edge weight, as if each particle
+    // in the cell is connected to all other particles in neighboring cell.
+    struct CountAndNode {
+      size_t count = 0;
+      size_t node = npos;
+    };
+    Mdvector<CountAndNode, Dim> cells{grid.num_cells().elems()};
+    par::for_each(points, [&grid, &cells](const auto& point) {
+      par::fetch_and_add(cells[grid.cell_index(point).elems()].count, 1);
+    });
+    size_t num_nodes = 0;
+    for (auto& [count, node] : cells) {
+      if (count > 0) node = num_nodes++;
+    }
+
+    // Build the graph connecting the cells.
+    /// @todo I wish this code to be cleaner. Once we have a proper graph
+    ///       library, we can simplify this code.
+    static constexpr auto MaxNumEdges = 2 * Dim;
+    using EdgeMap = InplaceFlatMap<size_t, size_t, MaxNumEdges>;
+    graph::CapWeightedGraph<MaxNumEdges> graph(num_nodes);
+    std::vector<graph::weight_t> node_weights(num_nodes);
+    par::for_each( //
```

Comment:
Let's find a way to make this more readable. Even by dropping parallelization here.

Reply:
No, that's not the right time...